### DISCUSSION TITLE
How should I replicate RSA_sign() with only highlevel calls?

### DISCUSSION THREAD

**Discussion by jjeffroberts** [openssl/openssl#22307] [Q&A]
Before `RSA_sign()` calls `RSA_private_encrypt()` it encodes the input data using `encode_pkcs1()`, which calls `ossl_rsa_digestinfo_encoding()` and makes use of various macros.

```
/*
 * Encodes a DigestInfo prefix of hash |type| and digest |m|, as
 * described in EMSA-PKCS1-v1_5-ENCODE, RFC 3447 section 9.2 step 2. This
 * encodes the DigestInfo (T and tLen) but does not add the padding.
 *
 * On success, it returns one and sets |*out| to a newly allocated buffer
 * containing the result and |*out_len| to its length. The caller must free
 * |*out| with OPENSSL_free(). Otherwise, it returns zero.
 */
static int encode_pkcs1(unsigned char **out, size_t *out_len, int type,
                        const unsigned char *m, size_t m_len)
```

I am updating my code to get rid of deprecated calls.

What is the best way to replace this part of `RSA_sign()`?

**Comment by paulidale** [comment]
I'd suggest starting with the [RSA PSS signing demo code](https://github.com/openssl/openssl/blob/master/demos/signature/rsa_pss_direct.c#L1):

https://github.com/openssl/openssl/blob/79997a919f6cf3823d04fa9b34adaaa5aadd871a/demos/signature/rsa_pss_direct.c#L38-L118

**Reply by jjeffroberts** [reply]
I'm pretty sure that was what I tried but it does not encode the data the same as RSA_sign() and I get a different result.

**Reply by paulidale** [reply]
You'll have to change the `EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING)` to match the padding mode you want.

**Reply by jjeffroberts** [reply]
Oh! That I have not tried yet.

**Comment by jjeffroberts** [comment]
I'm not asking the right question.
